plugins/zify/zify.ml

```diff
@@ -0,0 +1,1095 @@
+let get_type_of env evd e =
+  Tacred.cbv_beta env evd (Retyping.get_type_of env evd e)
+
+(** [unsafe_to_constr c] returns a [Constr.t] without considering an evar_map.
+    This is useful for calling Constr.hash *)
+let unsafe_to_constr = EConstr.Unsafe.to_constr
+
+let pr_constr env evd e =
+  Printer.pr_constr_env env evd (unsafe_to_constr e)
+
+
+let debug = false (** Debug flag to get some intermediate terms *)
+
+(** [get_arrow_typ evd t] returns [t1;.tn] such that t = t1 -> .. -> tn.ci_npar
+    (only syntactic matching)
+ *)
+let rec get_arrow_typ evd t  =
+  match EConstr.kind evd t with
+  | Prod (a,p1,p2) (*when a.Context.binder_name = Names.Anonymous*) ->
+       p1::(get_arrow_typ evd p2)
+  |  _ -> [t]
+
+
+(** [get_binary_arrow t] return t' such that t = t' -> t' -> t' *)
+let get_binary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with
+  | []  -> assert false
+  | [t1;t2;t3] ->  Some (t1,t2,t3)
+  |  _   -> None
+
+(** [get_unary_arrow t] return t' such that t = t' -> t' *)
+let get_unary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with
+  | []  -> assert false
+  | [t1;t2] -> Some (t1,t2)
+  |  _   -> None
+
+
+(** [HConstr] is a map indexed by EConstr.t.
+    It should only be used using closed terms.
+ *)
+module HConstr =
+  struct
+    module M =
+      Map.Make(
+          struct
+            type t = EConstr.t
+
+            let compare c c' = Constr.compare (unsafe_to_constr c) (unsafe_to_constr c')
+
+          end)
+
+    let lfind h m =
+      try M.find h m
+      with Not_found -> []
+
+
+    let add h e m =
+      let l = lfind h m in
+      M.add h (e::l) m
+
+    let empty = M.empty
+
+    let find h m =
+      match lfind h m with
+      | e::_ -> e
+      | [] -> raise Not_found
+
+
+    let find_all  = lfind
+
+    let fold f m acc =
+      M.fold (fun k l acc ->
+          List.fold_left (fun acc e -> f k e acc) acc l) m acc
+
+  end
+
+
+(** [get_projections_from_constant (evd,c) ]
+    returns an array of constr [| a1,.. an|] such that [c] is defined as
+    Definition c := mk a1 .. an with mk a constructor.
+    ai is therefore either a type parameter or a projection.
+*)
+let get_projections_from_constant (evd,i) =
+  match EConstr.kind evd i with
+  | Constr.Const(c,_ ) ->
+     begin
+       match Global.body_of_constant Library.indirect_accessor c with
+       | None -> failwith "Add Injection requires a constant (with a body)"
+       | Some(c,_,_) ->
+          match EConstr.kind evd  (EConstr.of_constr c) with
+          | App(c,a) -> Some a
+          |    _            -> None
+         end
+      |   _   -> None
+
+(** [reduce_term evd t] carefully unfolds aliases in [t] i.e.
+    if t = ctx[c] where c is a constant which
+    body is another constant c' then
+    [reduce_term env evd t] returns ctx[c']
+
+    This could be made more agressive..
+ *)
+
+let rec unfold_const evd c ctx =
+  match Global.body_of_constant Library.indirect_accessor c with
```

Comment:
Why are you opening opaques here?

Reply:
Search for an object of type  'Opaqueproof.indirect_accessor ', this is the only element I could get.
I am good with any suggestion.

Reply:
Don't use body_of_constant, use Environ.constant_opt_value_in instead.
Or better don't write some adhoc reduce_term and use `Reductionops.clos_norm_flags CClosure.delta` or some such thing

Reply:
Removed unfold_cst. Instead, for terms of the form @eq ty, ty is unified with candidates Z, positive etc.